horovod/common/ops/nccl_operations.cc

```diff
@@ -41,8 +41,9 @@ ncclDataType_t GetNCCLDataType(const std::shared_ptr<Tensor> tensor) {
   }
 }
 
-void NCCLContext::ErrorCheck(std::string op_name, ncclResult_t nccl_result) {
+void NCCLContext::ErrorCheck(std::string op_name, ncclResult_t nccl_result, ncclComm_t& nccl_comm) {
   if (nccl_result != ncclSuccess) {
+    ncclCommAbort(nccl_comm);
```

Comment:
As a comment, because NCCL ops are async, the error may not necessary be for this specific op but for a previous one: `this function may also return error codes from previous, asynchronous launches.`

Also, at the places we are waiting in our codebase for NCCL ops to finish, we probably need to call `ncclCommGetAsyncError()` to check for specific NCCL async errors related to network communications, etc... See https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/usage/communicators.html#asynchronous-errors-and-error-handling for code snippet.

Reply:
I'm putting together a follow-up PR to address the async error handling, in order to avoid making this PR any more complex.  Let me know if that sounds reasonable, and if the current PR looks good as a first step.